kube-controllers/tests/fv/metrics_test.go

```diff
@@ -318,88 +394,129 @@ var _ = Describe("kube-controllers metrics tests", func() {
+				`ipam_allocations_per_node{node="node-a"} 5`,
+				`ipam_allocations_per_node{node="node-c"} 2`,
+				`ipam_blocks{ippool="test-ippool-1",node="node-a"} 1`,
+				`ipam_blocks{ippool="test-ippool-1",node="node-b"} 1`,
+				`ipam_blocks{ippool="no_ippool",node="no_affinity"} 1`,
+				`ipam_blocks{ippool="test-ippool-1",node="no_affinity"} 1`,
+				`ipam_blocks_per_node{node="node-a"} 1`,
+				`ipam_blocks_per_node{node="node-b"} 1`,
+				`ipam_blocks_per_node{node="no_affinity"} 2`,
+				`ipam_allocations_borrowed{ippool="test-ippool-1",node="node-c"} 2`,
+				`ipam_allocations_borrowed_per_node{node="node-c"} 2`,
+			},
+			[]string{
+				`ipam_allocations_gc_reclamations{ippool="test-ippool-1",node="node-b"}`,
+				`ipam_allocations_in_use{ippool="test-ippool-1",node="node-b"}`,
+				`ipam_allocations_in_use{ippool="no_ippool",node="node-b"}`,
+				`ipam_allocations_per_node{node="node-b"}`,
+			},
+			kubeControllers.IP,
+			5*time.Second,
+		)
 
 		// Deleting NodeC should clean up the second and third blocks since both node B and C
 		// are now gone.
 		err = k8sClient.CoreV1().Nodes().Delete(context.Background(), nodeC, metav1.DeleteOptions{})
 		Expect(err).NotTo(HaveOccurred())
+		deletePodWithIP("pod-c", "192.168.0.129", k8sClient, calicoClient)
+		deletePodWithIP("pod-c2", "192.168.0.66", k8sClient, calicoClient)
+
 		Eventually(func() error {
 			if err := assertIPsWithHandle(calicoClient.IPAM(), handleC, 0); err != nil {
 				return err
 			}
 			if err := assertIPsWithHandle(calicoClient.IPAM(), handleC2, 0); err != nil {
 				return err
 			}
-			if err := assertNumBlocks(bc, 1); err != nil {
+			if err := assertNumBlocks(bc, 2); err != nil {
 				return err
 			}
 			return nil
 		}, time.Second*10, 500*time.Millisecond).Should(BeNil())
 
 		// Assert that IPAM metrics have been updated. Both blocks should now be released, and all that should be
 		// left is the block and allocations for node A.
-		notExpectedMetrics = []string{
-			`ipam_allocations_per_node{node="node-c"}`,
-			`ipam_allocations_per_node{node="node-b"}`,
-			`ipam_blocks_per_node{node="node-b"}`,
-			`ipam_blocks_per_node{node="node-c"}`,
-			`ipam_blocks_per_node{node="no_affinity"}`,
-			`ipam_allocations_borrowed_per_node{node="node-c"}`,
-		}
-		expectedMetrics = []string{
-			`ipam_allocations_per_node{node="node-a"} 4`,
-			`ipam_blocks_per_node{node="node-a"} 1`,
-		}
-		Eventually(func() error {
-			out, err = getMetrics(fmt.Sprintf("http://%s:9094/metrics", kubeControllers.IP))
-			Expect(err).NotTo(HaveOccurred())
-			for _, s := range expectedMetrics {
-				if !strings.Contains(out, s) {
-					return fmt.Errorf("Expected:\n%s\nTo contain metric:\n  %s\n", out, s)
-				}
-			}
-			for _, s := range notExpectedMetrics {
-				if strings.Contains(out, s) {
-					return fmt.Errorf("Expected:\n%s\nNOT to contain metric:\n  %s\n", out, s)
-				}
-			}
-			return nil
-		}, 5*time.Second).ShouldNot(HaveOccurred())
+		validateExpectedAndUnexpectedMetrics(
```

Comment:
could probably just call this `validateMetrics` given it's checking both expected AND unexpected :grin:

Reply:
I was verbose here intentionally so that it is clearer within the context of the test body what each string list is referring to (i.e. first list is expected, second is unexpected). Sometimes the IDE can help here by decorating the function all with the argument names, however I found GoLand wasn't doing that for these function calls :man_shrugging: